File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.js
function solution(board, skill) {
    const [R, C] = [board.length, board[0].length];
    const DP = Array.from({length : R}, () => Array(C).fill(0));
    for(const [type, r1, c1, r2, c2, degree] of skill) {
        const d = type === 1 ? -degree : degree;
        DP[r1][c1] += d;
        if(r2 + 1 < R) DP[r2 + 1][c1] -= d;
        if(c2 + 1 < C) DP[r1][c2 + 1] -= d;
        if(r2 + 1 < R && c2 + 1 < C) DP[r2 + 1][c2 + 1] += d;
    }
    let answer = 0;
    for(let i = 0; i < R; i++) {
        for(let j = 0; j < C; j++) {
            const top = i > 0 ? DP[i - 1][j] : 0;
            const left = j > 0 ? DP[i][j - 1] : 0;
            const intersection = i > 0 && j > 0 ? DP[i - 1][j - 1] : 0;
            DP[i][j] += top + left - intersection;
            if(DP[i][j] + board[i][j] > 0) answer++;
        }
    }
    return answer;
}

File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.py
def solution(board, skill):
    N, M = len(board), len(board[0])
    dp = [[0] * M for _ in range(N)]
    answer = 0
    
    for t, r1, c1, r2, c2, degree in skill:
        d = -degree if t == 1 else degree
        dp[r1][c1] += d
        if r2 + 1 < N:
            dp[r2 + 1][c1] -= d
        if c2 + 1 < M:
            dp[r1][c2 + 1] -= d
        if r2 + 1 < N and c2 + 1 < M:
            dp[r2 + 1][c2 + 1] += d
    
    for i in range(N): 
        for j in range(M):
            top = dp[i - 1][j] if i > 0 else 0
            left = dp[i][j - 1] if j > 0 else 0
            intersection = dp[i - 1][j - 1] if i > 0 and j > 0 else 0
            dp[i][j] += top + left - intersection 
            
            if dp[i][j] + board[i][j] > 0:
                answer += 1
    
    return answer
            